### PR TITLE
insertParagraph command creates additional empty paragraph when used at the end of block element

### PR DESCRIPTION
**Pull request by illia-stv**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Fix (paragraph): "insertParagraph" command doesn't insert two paragraphs when position is at the edge of the line. Closes [#13866](https://github.com/ckeditor/ckeditor5/issues/13866).

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-paragraph/tests/insertparagraphcommand.js
Comment (Dumluregn) [inline packages/ckeditor5-paragraph/tests/insertparagraphcommand.js:99]: `asd` :stuck\_out\_tongue\_closed\_eyes:

File: packages/ckeditor5-paragraph/src/insertparagraphcommand.ts
Comment (Dumluregn) [inline packages/ckeditor5-paragraph/src/insertparagraphcommand.ts:112]: There's already an API for it - [`Writer.createPositionAfter()`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_model_model-Model.html#function-createPositionAfter). I think we should be safe with `position =` _`writer`_`.createPositionAfter( position.parent as Item );`.

BTW - [https://dictionary.cambridge.org/dictionary/english/bellow](https://dictionary.cambridge.org/dictionary/english/bellow) :smile:

File: packages/ckeditor5-paragraph/src/insertparagraphcommand.ts
Comment (Dumluregn) [inline packages/ckeditor5-paragraph/src/insertparagraphcommand.ts:83]: All these conditions contain `position.path.length > 1`. Maybe it's possible to simplify them? What does it mean if it's equal to 0 or 1? Is an earlier condition possible?

File: packages/ckeditor5-paragraph/tests/insertparagraphcommand.js
Comment (Dumluregn) [inline packages/ckeditor5-paragraph/tests/insertparagraphcommand.js:124]: This test won't be necessary once you use the API I proposed.

**Review by Dumluregn — CHANGES_REQUESTED** [review]
So, the problem is that currently with the following structure: `<h2>[]</h2>`, `writer.split()` does exactly what we tell it - splits it into `<h2>[]</h2><h2></h2>`, and then we insert a paragraph in the middle (same applies for the `<p>` elements, but I used `<h2>` to make it clearer), ending up with two new blocks instead of one: `<h2></h2><p>[]</p><h2></h2>`. So I agree that we problem lies within the command, not `writer.split()` method.

The proposed solution makes sense. Nice that you've noticed that the issue occurs also at the beginning of the blocks, not just at the end. Let's clean it up a little and we'll be ready to pass the PR to the Core team.

File: packages/ckeditor5-paragraph/src/insertparagraphcommand.ts
Comment (mlewand) [inline packages/ckeditor5-paragraph/src/insertparagraphcommand.ts:106]:
```suggestion
export function createPositionBelow( writer: Writer, position: Position, allowedParent: Element ): Position {
```

You probably meant that 😉

File: packages/ckeditor5-paragraph/src/insertparagraphcommand.ts
Comment (illia-stv) [reply packages/ckeditor5-paragraph/src/insertparagraphcommand.ts:83]: If **path** contain only 1 value like [ 1 ], `getParentPath()` of that will be empty array ([ ]) then it's not valid path.

Comment (illia-stv) [reply packages/ckeditor5-paragraph/src/insertparagraphcommand.ts:83]: Eventually we could replace it with
`
if ( position.path.length < 2 ) {
				model.insertContent( paragraph, position );
				writer.setSelection( paragraph, 'in' );

				return;
			}

			if ( position.isAtStart && position.isAtEnd ) {
				// NOTE: In case when paragraph is empty.
				// <paragraph>[]</paragraph> ---> <paragraph></paragraph><paragraph>[]</paragraph>
				position = writer.createPositionAfter( position.parent as Item );
			} else if ( position.isAtStart && ( position.nodeAfter as Text ).data ) {
				// When position is at the start of the line we want to insert paragraph above.
				// <paragraph>[]foo</paragraph> ---> <paragraph>[]</paragraph><paragraph>foo</paragraph>
				position = writer.createPositionFromPath( allowedParent, position.getParentPath() );
			} else if ( position.isAtEnd && ( position.nodeBefore as Text ).data ) {
				// When position is at the end of the line we want to insert paragraph after.
				// <paragraph>foo[]</paragraph> ---> <paragraph>foo</paragraph><paragraph>[]</paragraph>
				position = writer.createPositionAfter( position.parent as Item );
			} else if ( !model.schema.checkChild( position.parent as Element, paragraph ) ) {
				// When position is inside the content we want to split it
				// <paragraph>fo[]o</paragraph>
				//            |
				//			  ↓
				// <paragraph>fo</paragraph>
				// <paragraph>[]</paragraph>
				// <paragraph>o</paragraph>
				position = writer.split( position, allowedParent ).position;
			}

			model.insertContent( paragraph, position );

			writer.setSelection( paragraph, 'in' );
`

**Review by Dumluregn — APPROVED** [review]
LGTM, waiting for CI.

**Review by arkflpc — APPROVED** [review]

File: packages/ckeditor5-paragraph/src/insertparagraphcommand.ts
Comment (niegowski) [inline packages/ckeditor5-paragraph/src/insertparagraphcommand.ts:91]: What about inline elements? For example inline image? Maybe you could check `position.parent.isEmpty` instead of checking `nodeBefore`/`nodeAfter`?

Comment (illia-stv) [reply packages/ckeditor5-paragraph/src/insertparagraphcommand.ts:91]: We decided to keep previous behaviour for **non-text** elements, look at this test: https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-widget/tests/widgettypearound/widgettypearound.js#L1056-L1089
here this kind behaviour is expected for **widgets** then we fix issue only for **text** nodes

Comment (niegowski) [reply packages/ckeditor5-paragraph/src/insertparagraphcommand.ts:91]: I'm talking about inline widgets, not block widgets. They are allowed everywhere text is allowed.

**Review by Dumluregn — APPROVED** [review]

**Pull request merged**
